### PR TITLE
Add separate log for each line in batch run.

### PR DESCRIPTION
**Pull request by huaiyan**
# Description
Add separate log for each line in batch run. Which could improve user experience for debugging failed lines.
In future, it could also be useful to check eager flow's customer print content.

Local batch run flow logs folder example:
image
Use fixed length file name to keep right order, then it could be easier to check.

Implementation:
There are 3 loggers in executor ("execution"\"execution.flow"\"execution.bulk")
They all have handler list to write log.
Such as `FileHandlerConcurrentWrapper` and `StreamHandler`.
`FileHandlerConcurrentWrapper` is wrapper for `FileHandler`.
Before this PR, we use `FileHandlerConcurrentWrapper` to write bulk run's log content to same file.
In this PR, we create a new `FileHandlerConcurrentWrapper` for "execution" and "execution.flow", to write their log content to different file according to line number.

So, for each line execution, we should set log context after getting `line_number` info. This part of logic is implemented by changing invocation of `log_context_initialization_func`

This pull request introduces changes to the logging mechanism in the PromptFlow SDK, specifically for batch runs. The changes allow for line-specific logging in batch runs, with each line's log stored in a separate file in a specified folder. The `LocalStorageFilenames` class and `LoggerOperations` class have been modified to accommodate the new `line_log_folder` attribute. A new class, `LineFileHandlerConcurrentWrapper`, has been introduced to ensure thread safety when writing to different log files in different contexts. The `get_logger` function now adds an instance of this new class to the logger. The `LogContext` class has been extended with `line_log_folder` and `line_number` attributes. The `_set_log_path` method in `LogContext` has been modified to set the path for the line log file if the run mode is `Batch`. Finally, the `_exec_line_for_queue` function in the `LineExecutionProcessPool` class has been updated to set the logger context for each line execution.

Main changes:

Updates to logging:

* [`src/promptflow/promptflow/_sdk/_constants.py`](diffhunk://#diff-9ffef684cefef289f51e58b0aff1cd6cfa42363ace309d07f36dfe6559c5b84dR287): Added `LINE_LOG_FOLDER` to `LocalStorageFilenames` class.
* [`src/promptflow/promptflow/_sdk/operations/_local_storage_operations.py`](diffhunk://#diff-1a5c2463c2ef997c5203a6b2e77b247d812961b7a6df5f0f45c02579b6656719R91-R100): Added `line_log_folder` attribute to `get_initializer` method and `LoggerOperations` class. [[1]](diffhunk://#diff-1a5c2463c2ef997c5203a6b2e77b247d812961b7a6df5f0f45c02579b6656719R91-R100) [[2]](diffhunk://#diff-1a5c2463c2ef997c5203a6b2e77b247d812961b7a6df5f0f45c02579b6656719L199-R205)
* [`src/promptflow/promptflow/_utils/logger_utils.py`](diffhunk://#diff-4cf50231dfaf296f999ac6c39e74c09399635d8aed56e609d5fdee0e63574b87R171-R182): Introduced `LineFileHandlerConcurrentWrapper` class for thread safety, added it to the `get_logger` function, and added `line_log_folder` and `line_number` attributes to `LogContext` class. Modified `_set_log_path` method to set the path for line log file in batch mode. [[1]](diffhunk://#diff-4cf50231dfaf296f999ac6c39e74c09399635d8aed56e609d5fdee0e63574b87R171-R182) [[2]](diffhunk://#diff-4cf50231dfaf296f999ac6c39e74c09399635d8aed56e609d5fdee0e63574b87R199) [[3]](diffhunk://#diff-4cf50231dfaf296f999ac6c39e74c09399635d8aed56e609d5fdee0e63574b87R233-R242) [[4]](diffhunk://#diff-4cf50231dfaf296f999ac6c39e74c09399635d8aed56e609d5fdee0e63574b87R294-R304) [[5]](diffhunk://#diff-4cf50231dfaf296f999ac6c39e74c09399635d8aed56e609d5fdee0e63574b87R336-R340)
* [`src/promptflow/promptflow/executor/_line_execution_process_pool.py`](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L636-R636): Updated `_exec_line_for_queue` function to set logger context for each line execution. [[1]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L636-R636) [[2]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1R680-R692)

Minor changes:

* [`src/promptflow/promptflow/executor/_line_execution_process_pool.py`](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L19-R19): Imported `Callable` from `typing`.

# All Promptflow Contribution checklist:
- [X] **The pull request does not introduce [breaking changes].**
- [X] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by zhengfeiwang — DISMISSED** [review]

File: src/promptflow/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
   - CLI: Support `pfazure run create --resume-from <original-run-name>` to create a run resume from another run.
   - SDK: Support `pf.run(resume_from=<original-run-name>)` to create a run resume from another run.
 
+- [Executor] Added per-line logging for batch runs, stored under the `flow_logs` folder.
```
Comment (thy09) [inline src/promptflow/CHANGELOG.md:14]: Maybe `[Batch]` is more suitable instead of [Executor]

Comment (huaiyan) [reply src/promptflow/CHANGELOG.md]: Fixed.

**Review by zhengfeiwang — APPROVED** [review]

File: src/promptflow/promptflow/_utils/logger_utils.py [hunk 2]
```diff
@@ -271,6 +278,18 @@ def _set_log_path(self):
                     handler = FileHandler(self.file_path)
                     log_handler.handler = handler
 
+    def _set_batch_run_flow_logs_handler(self):
```
Comment (thy09) [inline src/promptflow/promptflow/_utils/logger_utils.py:281]: I think we need a description how we implement this feature, it is hard to understand the code here.

File: src/promptflow/promptflow/_utils/logger_utils.py [hunk 3]
```diff
@@ -271,6 +278,18 @@ def _set_log_path(self):
...
+        if self.run_mode != RunMode.Batch or self.flow_logs_folder is None or self.line_number is None:
+            return
+        from pathlib import Path
```
Comment (thy09) [inline src/promptflow/promptflow/_utils/logger_utils.py:284]: Do not import here.

Comment (huaiyan) [reply src/promptflow/promptflow/_utils/logger_utils.py:284]: Fixed.

File: src/promptflow/promptflow/_utils/logger_utils.py (hunk 2 above)
Comment (huaiyan) [reply src/promptflow/promptflow/_utils/logger_utils.py]: Added. And don't create `FileHandlerConcurrentWrapper` for each __enter__ method

**Review by thy09 — APPROVED** [review]

**Review by zhengfeiwang — APPROVED** [review]

**Pull request closed without merging**